File: hash/프로그래머스lv1-완주하지못한선수.js
// completion 기준으로 dictionary를 만든다.

function solution(participant, completion) {
  let obj = {};
  completion.forEach((v, i) => {
    if (obj[v]) obj[v]++;
    else obj[v] = 1;
  });
  // 여기서 forEach를 쓰면 안된다. forEach는 undefined를 리턴하기 때문에
  // const result = map... 도 좋지 않다. 모든 구간에서 return 하지 않은 값들에 대해 undefined를 리턴하기 때문에 [undefined, undefined, undefined ... 정답, undefined ...] 형태가 되버림
  for (let i = 0; i < participant.length; i++) {
    const x = participant[i];
    if (!obj[x]) return x;
    obj[x]--;
  }
}

/*
정확성  테스트
테스트 1 〉	통과 (0.23ms, 30.4MB)
테스트 2 〉	통과 (0.23ms, 30.3MB)
테스트 3 〉	통과 (0.29ms, 30.3MB)
테스트 4 〉	통과 (0.32ms, 30.5MB)
테스트 5 〉	통과 (0.40ms, 30.3MB)
효율성  테스트
테스트 1 〉	통과 (16.40ms, 46.9MB)
테스트 2 〉	통과 (19.73ms, 51.1MB)
테스트 3 〉	통과 (21.42ms, 54.9MB)
테스트 4 〉	통과 (28.08ms, 62.4MB)
테스트 5 〉	통과 (33.24ms, 62.6MB)
*/

// 단순 sorting
function solution(participant, completion) {
  let par = participant.sort();
  let com = completion.sort();
  let i = 0;
  while (true) {
    i++;
    if (par[i] !== com[i]) {
      break;
    }
  }
  return par[i];
}
/*
정확성  테스트
테스트 1 〉	통과 (0.07ms, 30MB)
테스트 2 〉	통과 (0.07ms, 30.1MB)
테스트 3 〉	통과 (0.45ms, 30.2MB)
테스트 4 〉	통과 (0.80ms, 30.3MB)
테스트 5 〉	통과 (0.59ms, 30.1MB)
효율성  테스트
테스트 1 〉	통과 (43.82ms, 41.2MB)
테스트 2 〉	통과 (75.16ms, 47.8MB)
테스트 3 〉	통과 (95.45ms, 52.7MB)
테스트 4 〉	통과 (105.75ms, 55.3MB)
테스트 5 〉	통과 (104.44ms, 53.8MB)
*/

File: hash/프로그래머스lv1-완주하지 못한 선수.py
from collections import defaultdict
def solution(participant, completion):
    table = defaultdict(int)
    
    for person in completion:
        table[person] += 1
    
    for person in participant:
        if person not in table or table[person] == 0:
            return person
        table[person] -= 1

"""
정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.3MB)
테스트 2 〉	통과 (0.01ms, 10.2MB)
테스트 3 〉	통과 (0.14ms, 10.1MB)
테스트 4 〉	통과 (0.35ms, 10.3MB)
테스트 5 〉	통과 (0.28ms, 10.3MB)
효율성  테스트
테스트 1 〉	통과 (19.51ms, 21.7MB)
테스트 2 〉	통과 (35.41ms, 25.1MB)
테스트 3 〉	통과 (37.48ms, 27.5MB)
테스트 4 〉	통과 (38.35ms, 34MB)
테스트 5 〉	통과 (60.18ms, 33.9MB)
"""

# Counter 사용
from collections import Counter
def solution(participant, completion):
    part = Counter(participant)
    com = Counter(completion)
    
    for key in part.keys():
        if not com[key]:
            return key
        com[key] -= part[key]
        if com[key] < 0:
            return key

"""
정확성  테스트
테스트 1 〉	통과 (0.03ms, 10.2MB)
테스트 2 〉	통과 (0.04ms, 10.2MB)
테스트 3 〉	통과 (0.18ms, 10.4MB)
테스트 4 〉	통과 (0.67ms, 10.4MB)
테스트 5 〉	통과 (0.28ms, 10.4MB)
효율성  테스트
테스트 1 〉	통과 (20.80ms, 24.4MB)
테스트 2 〉	통과 (42.82ms, 27.8MB)
테스트 3 〉	통과 (41.13ms, 30.2MB)
테스트 4 〉	통과 (45.21ms, 39MB)
테스트 5 〉	통과 (31.45ms, 38.9MB)
"""